modules/fedimint-ln-client/src/db.rs

```diff
@@ -80,3 +87,75 @@ impl_db_record!(
 );
 
 impl_db_lookup!(key = MetaOverridesKey, query_prefix = MetaOverridesPrefix);
+
+/// DB migration from version 0 to version 1
+pub(crate) async fn migrate_to_v1(
+    module_instance_id: ModuleInstanceId,
+    active_states: Vec<(Vec<u8>, OperationId)>,
+    inactive_states: Vec<(Vec<u8>, OperationId)>,
+    decoders: ModuleDecoderRegistry,
+) -> anyhow::Result<Option<(Vec<DynState>, Vec<DynState>)>> {
+    let mut new_active_states = Vec::with_capacity(active_states.len());
+    for (active_state, _) in active_states {
+        // Try to decode the bytes as a `DynState`
+        let dynstate = DynState::from_bytes(active_state.as_slice(), &decoders)?;
```

Comment:
To not require a complete, old copy of `LightningClientStateMachines` I'd implement some manual decoding here.

Since we are only interested in the old `SubmittedOffer` state the byte sequence I'd be expecting here is (with concrete values we can match added after `=`):

```
<module_instance_id:varint> <ln_sm_variant:varint=2> <ln_sm_len:varint> <operation_id:32b> <receive_sm_variant:varint=0> <receive_sm_len:varint> <LightningReceiveSubmittedOfferV0>
```

If you parse all SMs that match this pattern using the old `LightningReceiveSubmittedOfferV0` struct while parsing everything else directly as a `DynState` you can implement the migration logic without keeping many redundant structs around while also avoiding polluting production enums with no-longer-used variants.

Reply:
Have this working, however haven't removed the enum variant because otherwise I am unsure on how to do construct the tests. `create_client_states` needs to return a `Vec<LightningClientStateMachines>`, not sure how we'd do that without keeping the `SubmittedOfferV0` variant.